reference/opt/shaders-msl/asm/tesc/tess-level-overrun.multi-patch.asm.tesc

```diff
@@ -0,0 +1,26 @@
+#include <metal_stdlib>
+#include <simd/simd.h>
+
+using namespace metal;
+
+struct TessLevels
+{
+    float inner0;
+    float inner1;
+    float outer0;
+    float outer1;
+    float outer2;
+    float outer3;
+};
+
+kernel void main0(const device TessLevels& sb_levels [[buffer(0)]], uint3 gl_GlobalInvocationID [[thread_position_in_grid]], constant uint* spvIndirectParams [[buffer(29)]], device MTLTriangleTessellationFactorsHalf* spvTessLevel [[buffer(26)]])
+{
+    if (gl_GlobalInvocationID.x / 1 >= spvIndirectParams[1])
+        return;
```

Comment:
We always had the potential issue of barrier in tesc after returning, which should be undefined. Do we avoid that problem if partial workgroups are supported?

Reply:
Unfortunately, we can't. I wanted to add a knob that omits this check if non-uniform workgroups are supported, but this doesn't work for indirect draws: there's no `-[MTLComputeCommandEncoder dispatchThreadsWithIndirectBuffer:...]` method.

Reply:
Come to think of it, I have been seeing weirdness in the `dEQP-VK.tessellation.shader_input_output.barrier` test. The geometry isn't offset correctly, because the per-vertex varyings aren't written correctly in the `tesc` shader. There was a bump at 4-5 which was closer to what it should be, which led me to suspect that the driver was miscompiling the shader around a conditional branch for vertex 5. But this might explain some of the weirdness.

Reply:
I've wondered if, like in Apple's example ported hull shader, I should just wrap the whole shader in an `if` statement. But then the shaders that aren't supposed to be running won't hit the barriers, which will cause the other shaders to hang.

Reply:
Not sure that would help. If we know that the workgroup is <= SIMD width, it might happen to work though. I think we need to have a solution for this problem in a new implementation though.

Reply:
Since number of control points is <= 32, if SIMD width >= 32, we should be able to use 64-wide workgroups on AMD as well with simd barriers.

Reply:
MSL 1.2. Oh, and SIMD-group barriers require MSL 2.0 on Mac.

Reply:
There's nothing in the MSL spec to suggest that barriers are undefined after a return statement. It only says:

> If `threadgroup_barrier` (or `simdgroup_barrier`) is inside a conditional statement and if any thread enters the conditional statement and executes the barrier function, then all threads in the threadgroup (or SIMD-group) must enter the conditional and execute the barrier function.
>
> If `threadgroup_barrier` (or `simdgroup_barrier`) is inside a loop, for each iteration of the loop, all threads in the threadgroup (or SIMD-group) must execute the barrier function before any threads continue execution beyond the barrier function.

I wonder if the behavior here is under-documented.

Reply:
I would be surprised if it has different semantics than Vulkan or D3D here, "all threads must enter" means that any returned threads cannot enter and that is undefined behavior in those APIs. If the semantics are actually to make this well-behaved in Metal, then that is a very pleasant surprise. Needs to be well tested though.

Reply:
Oh, and the other thing about SIMD-groups, which I mentioned on #719. The driver is allowed to vary the SIMD-group size based on register usage--like in Vulkan's `VK_EXT_subgroup_size_control` extension, only it can't be turned off. So I can't necessarily guarantee that the SIMD-group size will be >= the number of control points.

Reply:
Is it possible to query the minimum SIMD-width though?

Reply:
No, not even with any of the secret undocumented methods.

Reply:
All right, I've made a test program which will reproduce the problem with the test. It replicates roughly what MoltenVK does in response to the test's actions. As you can see, I commented out the return, and... it had no effect. The output is still wrong. (You can see the output by running it under Xcode's control. It will automatically trigger a GPU capture with no action on your part.)

(Gaah! GitHub is being dumb. It won't let me upload the file. Sorry, I can't share it with you. You'll just have to take my word for it.)

Reply:
One test to try is using a large workgroup and then do this early return, doesn't have to be tess, just some compute kernel. I figure that for AMD you'll see 1 SIMD group, and I doubt s_barrier will care in that case, but it might for Intel and iOS where you usually get much smaller SIMD groups.

Reply:
On AMD, it works fine, even if some of the threads return early.

On Intel, I saw hangs.

Well that sucks. I can't have the extra threads return early apparently, but I can't have them actually write anything, or they might write out of bounds and cause a GPU page fault. What do I do now?

Reply:
Ouch, this was expected, but unfortunate. I think in 99.9% of cases, barriers() happen outside of control flow in hull shaders. In that case it should be possible to do a transform like:

```
void main()
{
    bool active_lane = ?;
    if (active_lane) { do_stuff(); }
    barrier();
    if (active_lane) { do_other_stuff(); }
}
```
However this assumption breaks down if the app somehow emits barriers in control flow. HLSL will never do this, and GLSL bans it, but SPIR-V technically supports it I think (not sure though! It might be our saving grace if not).

Another approach is padding. I think we can make it work just fine if we allocate padded space for the intermediate output buffers. To avoid problems with unintentional out-of-bounds reads we can clamp the reported gl_PrimitiveID to the application. That way we get duplicated work at the tail end of the dispatch. We also have to ban side-effects in tessellation so that application cannot observe duplicated execution of the tessellation control shader (it's a separate feature bit), which shouldn't be a big issue.

Reply:
Hi, I'm late to the party and catching up with posts and emails after my vacation.
We had an issue in a compute kernel where a barrier blocked the entire GPU because some threads had an early exit path. We filed an issue at Apple and some days later they reported the driver is fine, our compute kernel just fall into undefined behavior because of the conditional barriers.

Reply:
I've decided to just remove the boundary checks and have MoltenVK pad the buffers. Adding boundary checks around everything except barriers just seemed too complex.